cmd/autoscaler/main.go

```diff
@@ -159,14 +165,74 @@ func main() {
 		logger.Fatalw("Failed to start informers", zap.Error(err))
 	}
 
-	go controller.StartAll(ctx, controllers...)
+	// Needs to be called after informaers have started.
+	ss, err := statefulsetLister.StatefulSets(ns).Get(statefulSetName)
+	if err != nil {
+		logger.Fatalw("Failed to get autoscaler StatefulSet", zap.Error(err))
+	}
 
-	go func() {
+	acceptSm := func(sm asmetrics.StatMessage) {
+		collector.Record(sm.Key, time.Now(), sm.Stat)
+		multiScaler.Poke(sm.Key, sm.Stat)
+	}
+	processSms := func() {
 		for sm := range statsCh {
-			collector.Record(sm.Key, time.Now(), sm.Stat)
-			multiScaler.Poke(sm.Key, sm.Stat)
+			acceptSm(sm)
+		}
+	}
+
+	bucketSize := int(*ss.Spec.Replicas)
+	if bucketSize > 0 {
+		// If there are multiple pods for autoscaler, they must be run in StatefulSet
+		// ordinal assignment leader electior mode. So enable it.
+		ctx = leaderelection.WithDynamicLeaderElectorBuilder(
+			ctx, kubeClient, leaderelection.ComponentConfig{Buckets: uint32(bucketSize)})
+
+		// Build stats forwarding network.
+		bt, err := leaderelection.BuildBucketSet(bucketSize)
+		if err != nil {
+			logger.Fatalw("Failed to build bucket set", zap.Error(err))
 		}
-	}()
+		ordinal, err := leaderelection.ControllerOrdinal()
+		if err != nil {
+			logger.Fatalw("Failed to get ordinal of this pod", zap.Error(err))
+		}
+
+		wss := map[int]*websocket.ManagedConnection{}
+		for i := 0; i < bucketSize; i++ {
+			if i == ordinal {
+				continue
+			}
+			bkt, err := bt.Bucket(i)
+			if err != nil {
+				logger.Fatalf("Failed to get bucket with ordinal %d", i)
+			}
+			logger.Info("Connecting to Autoscaler at ", bkt.Name())
+			statSink := websocket.NewDurableSendingConnection(bkt.Name(), logger)
+			defer statSink.Shutdown()
+			wss[i] = statSink
+		}
+
+		processSms = func() {
+			for sm := range statsCh {
+				b, err := bt.BucketOrdinal(sm.Key)
+				if err != nil {
+					logger.Warnf("failed to fetch bucket ordinal for key %v: %v", sm.Key, err)
+					continue
+				}
+
+				if b != ordinal {
+					wss[b].Send(sm)
+				} else {
+					acceptSm(sm)
+				}
+			}
+		}
+	}
```

Comment:
Can we extract this into a few  functions.

Reply:
done